src/server/services/wfs/qgswfsutils.cpp

```diff
@@ -153,6 +153,47 @@ namespace QgsWfs
       request.setFlags( QgsFeatureRequest::ExactIntersect | QgsFeatureRequest::NoFlags );
       return request;
     }
+    // Apply BBOX through filterRect even inside an And to use spatial index
+    else if ( filterElem.firstChildElement().tagName() == QLatin1String( "And" ) &&
+              !filterElem.firstChildElement().firstChildElement( QLatin1String( "BBOX" ) ).isNull() )
+    {
+      QDomElement childElem = filterElem.firstChildElement().firstChildElement();
+      while ( !childElem.isNull() )
+      {
+        QDomElement childFilterElement = filterElem.ownerDocument().createElement( QLatin1String( "Filter" ) );
+        childFilterElement.appendChild( childElem.cloneNode( true ) );
+        QgsFeatureRequest childRequest = parseFilterElement( typeName, childFilterElement );
+        if ( childElem.tagName() == QLatin1String( "BBOX" ) )
+        {
+          if ( request.filterRect().isEmpty() )
+          {
+            request.setFilterRect( childRequest.filterRect() );
+          }
+          else
+          {
+            request.setFilterRect( request.filterRect().intersect( &childRequest.filterRect() ) );
+          }
+        }
+        else
+        {
+          if ( !bool( request.filterExpression() ) )
```

Comment:
Can you rewrite this to `request.filterExpression().isEmpty()`?

Reply:
Ah wait, this returns a `QgsExpression*`, not a string. Just drop the `bool()` in this case.